runtime/gc_modron_startup/mgcalloc.cpp

```diff
@@ -454,11 +456,17 @@ J9AllocateObject(J9VMThread *vmThread, J9Class *clazz, uintptr_t allocateFlags)
 				objectPtr, 
 				sizeInBytesRequired);
 		} else {
-			TRIGGER_J9HOOK_VM_OBJECT_ALLOCATE(
-				vmThread->javaVM->hookInterface, 
-				vmThread,
-				objectPtr, 
-				sizeInBytesRequired);
+			if (J9_EVENT_IS_HOOKED(vmThread->javaVM->hookInterface, J9HOOK_VM_OBJECT_ALLOCATE)) {
+				/* The JIT optimization only uses instrumentable allocate, so clear the NOT_AT_SAFE_POINT
+				 * bit now to allow the hook to run with no restrictions.
+				 */
+				VM_VMAccess::clearPublicFlags(vmThread, J9_PUBLIC_FLAGS_NOT_AT_SAFE_POINT);
```

Comment:
So, if flag is cleared here we continue to execute this function to the end with this flag cleared (and clear it again at the very end). This is confusing asymmetry.

Reply:
Agreed, but it doesn't matter in practice - the NOT_AT_SAFE_POINT serves two purposes - making GC have priority (flag set around the allocate operation itself), and enabling the JIT optimization. The JIT only uses instrumentable allocate, so this path does not matter to the JIT. Unsetting the flag allows the non-instrumentable allocate event to function properly (avoid asserting due to NOT_AT_SAFE_POINT being set during normal execution).

Reply:
I originally only cleared the bit once (recorded whether it happened early), but that was extra code which didn't really accomplish anything useful.